hashintel/hash
# Reimplement `queryEntities` with transpilation to structural queries

## 🌟 What is the purpose of this PR?

As part of experimenting with the Type System in HASH, we removed the `aggregateEntities` resolvers and implemented a stand-in `getAllLatestEntities`. This reimplements the (now called `queryEntities`) endpoint to match the updates made in the `0.3` release of the Block Protocol specification.

There are some operations which are currently unsupported but will be in the near(ish) future.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/0/1203749694686234/f) _(internal)_

## 🚫 Blocked by

N/A

## 🔍 What does this change?

- Implements the resolver for `queryEntities` which takes the BP aggregation syntax and translates it into a structural query across the HASH graph

## 📜 Does this require a change to the docs?

Not presently

## ⚠️ Known issues

Quite a few deficiencies were discovered, please see the added TODOs in the code 

## 🐾 Next steps

- Expand structural querying to support the deficiencies identified
- ~~Update the aggregate syntax in the BP to take a list of fields instead of a single one~~

## 🛡 What tests cover this?

Unsure if there are any besides code analysis

## ❓ How to test this?

1.  Checkout the branch / view the deployment
1.  Go to the GraphQL endpoint
1.  Confirm that the `aggregateEntities` query works as expected, for example:
```graphql
query {
  queryEntities(
    operation: {
      multiFilter:{
        filters: [
          {
            field: ["metadata", "entityTypeId"]
            operator:"EQUALS"
            value: "http://localhost:3000/@system-user/types/entity-type/page/v/1"
          }
        ]
        operator: "AND"
      }
    }
    constrainsLinksOn: {outgoing:0}
		constrainsValuesOn:{outgoing:0}
    constrainsPropertiesOn:{outgoing:0}
    constrainsLinkDestinationsOn:{outgoing:0}
    isOfType:{outgoing:0}
    hasLeftEntity:{outgoing:0,incoming:0}
    hasRightEntity:{outgoing:0,incoming:0}
  ) {
    roots
    vertices
    edges
  }
}
```

or

```graphql
query {
  queryEntities(
    operation: {
      multiFilter:{
        filters: [
          {
            field: ["properties", "http://localhost:3000/@system-user/types/property-type/title/"]
            operator: "EQUALS"
            value: "Middle"
          }
          {
            field: ["properties", "http://localhost:3000/@system-user/types/property-type/title/"]
            operator: "EQUALS"
            value: "First"
          }
        ]
        operator: "OR"
      }
    }
    constrainsLinksOn: {outgoing:0}
		constrainsValuesOn:{outgoing:0}
    constrainsPropertiesOn:{outgoing:0}
    constrainsLinkDestinationsOn:{outgoing:0}
    isOfType:{outgoing:0}
    hasLeftEntity:{outgoing:0,incoming:0}
    hasRightEntity:{outgoing:0,incoming:0}
  ) {
    roots
    vertices
    edges
  }
}
```

## 📹 Demo
<img width="1827" alt="image" src="https://user-images.githubusercontent.com/25749103/235985376-b45004b9-11d9-42b9-a16d-9eced0792e13.png">
